0000-scheduled-circuit.md

```diff
@@ -0,0 +1,266 @@
+
+
+Qubit = int # assume only physical qubits
+ArgumentedInstruction = Tuple[Instruction, List[Qubit], List[Clbit]]
+Duration = float
+LengthedInstruction = Tuple[ArgumentedInstruction, Duration]
+LengthedInstructions = Union[LengthedInstruction, Iterable[LengthedInstruction]]
+StartTime = float
+TimedInstruction = Tuple[StartTime, LengthedInstruction]
+Index = int
+Position = Tuple[Qubit, Index]
+Positions = Union[Position, Iterable[Position]]
+
+
+class ScheduledCircuit:
+    """Circuit with timings for each instruction."""
+
+    def __init__(self, name=None):
+        # TBD
+
+    def append(self,
+               target: LengthedInstruction,
+               buffer_duration: Duration = 0):
+        """Mutably append the target instruction after the latest instruction on its acting qubits.
+        For all the acting qubits, time-slots are filled with delay/idle instructions until the end time of the latest one.
+        If `buffer_duration` > 0, additional delay/idle instruction is also appended before the target instruction.
+        """
+
+    def instructions(self) -> List[LengthedInstruction]:
+        """Instructions sorted by their start time."""
+
+    def makespan(self, *qubits: int):
+        """The lastest end time of instructions on the qubits."""
+
+    def filter(self,
+               instruction_type: Instruction,
+               *qubits: Qubit,
+               unique: bool = True) -> List[LengthedInstruction?Position?]:
+        """Return instructions with only the instructions from this ``ScheduledCircuit`` which pass
+        though the provided properties; i.e. an instruction will be retained iff
+        the instruction type is contained in ``instruction_types`` and
+        occurs on a qubit contained in ``qubits`` (No qubits means *all* qubits).
+        If unique is False, multi-qubit instrucsions occurs multiple times in the result.
+        """
+
+    def replace(self,
+                reference: LengthedInstructions?Positions?,
+                *targets: LengthedInstructions?Positions?):
+        """Mutably replace the reference instruction just before the reference instruction.
+        Raise CircuitError if total duration of the targets does not equals to the duration of reference.
+        """
+
+    # """Ordered instructions on a specified qubit."""
+    # def instructions_on(self, qubit_index: int) -> List[LengthedInstruction]:
+
+
+def build_schedule(timed_circuit: ScheduledCircuit,
+                   inst_map: InstructionScheduleMap,
+                   ...) -> pulse.Schedule:
+    """Build schedule by replacing instructions in timed_circuit with schedules defined in inst_map.
+    """
+```
+
+## Alternative Approaches
+### float time vs int time:
+- float (seconds) = gate time in BackendProperties
+- int (number of cycle time) = pulse duration is defined by samples with dt or dtm
+  - -> float time for now (convert float time into int time in pulse.build_schedule())
```

Comment:
I do not think floating-point is a good idea. This will likely result in many difficulties in the actual implementation as there will be floating-point comparisons everywhere which will have many subtleties and be costly from a computational perspective. This is asking for many many headaches of having to check `assert_almost_equal` etc. when scheduling. With an integer-based formulation this calculation will have to be done only once, without comparison in Aer for the calculation of dissipators.

Furthermore, it will make moving towards the notion of a cycle much more difficult which is very crucial from a long term perspective as we introduce control flow. This also will increase the non-uniformity between pulse and circuits, when we should be trying to move towards more uniformity.